### PR TITLE
Fullscreen shell example

### PR DESCRIPTION
**Pull request by 4tuw2tuv0**
I think this example only takes fullscreen so I can't run it inside of an already-running Wayland instance (at least, I don't think) so I can't test it, in full, but I think, even if there's anything which still needs finishing, I think there's much more which is properly finished than not. The places where I'm uncertain, I'll leave comments.

I put comments throughout the example to the line number from the original C file so it's easier for new users to compare what happened in the original with the `guile-wlroots` version.

### CODE REVIEW

File: wlroots/types/fullscreen-shell-v1.scm [hunk 1]
```diff
@@ -0,0 +1,33 @@
+(define-module (wlroots types fullscreen-shell-v1)
```
Comment (Z572) [inline wlroots/types/fullscreen-shell-v1.scm]: please remove v1, different version should in same module，user should known what version need, so just import `(wlroots types fullscreen-shell)` and use expect version procedure

File: wlroots/types/fullscreen-shell-v1.scm [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  #:use-module (wayland display)
+  #:use-module (wlroots types)
+  #:use-module (wlroots utils)
+  #:export (wlr-fullscreen-shell-v1-create
+            .data
+            .display-destroy
+            .global
+
+            wlr-fullscreen-shell-v1-present-surface-event
```
Comment (Z572) [inline wlroots/types/fullscreen-shell-v1.scm]: should `<wlr-fullscreen-shell-v1-present-surface-event> `, define-wlr-types-class auto add `<` `>`

File: wlroots/types/output.scm [hunk 3]
```diff
@@ -197,6 +197,16 @@
...
 
+(define-wlr-procedure (wlr-output-effective-resolution output)
+  (ffi:void "wlr_output_effective_resolution" '(* * *))
+  (let ((width (bytestructure int))
+        (height (bytestructure int)))
+    (% (unwrap-wlr-output output)
+       (bytestructure->pointer width)
+       (bytestructure->pointer height))
+
+    (values (bytestructure-ref width) (bytestructure-ref height))))
```
Comment (Z572) [inline wlroots/types/output.scm]: i think should use `cons`, just get a width is not usually

File: wlroots/render/renderer.scm [hunk 4]
```diff
@@ -36,6 +38,17 @@
   (ffi:int "wlr_renderer_clear" `(* *))
   (% (unwrap-wlr-renderer renderer) (color->pointer color)))
 
+(define-wlr-procedure (wlr-render-texture-with-matrix r texture matrix alpha)
+  (ffi:int8 "wlr_render_texture_with_matrix" (list '* '* '* ffi:float))
+  (not (zero? (% (unwrap-wlr-renderer r)
+                 (unwrap-wlr-texture texture)
+                 (if (>= (length matrix) 9)
```
Comment (Z572) [inline wlroots/render/renderer.scm]: i don't think > 9 is usually,maybe just 9.

File: wlroots/render/renderer.scm [hunk 5]
```diff
@@ -36,6 +38,17 @@
...
+                     (bytestructure->pointer
+                       (bytestructure (bs:vector (length matrix) float)
+                                      (list->vector matrix)))
+                   (error "warning: list argument is too small; matrix requires, at least, 9"))
```
Comment (Z572) [inline wlroots/render/renderer.scm]: don't use `warning: ` begin, and use `srfi-34` `raise` like
`(raise (condition
                (&message
                 (message  "Not a 9 length vector or list"))))`

File: wlroots/render/renderer.scm [hunk 6]
```diff
@@ -36,6 +38,17 @@
...
   (% (unwrap-wlr-renderer renderer) (color->pointer color)))
 
+(define-wlr-procedure (wlr-render-texture-with-matrix r texture matrix alpha)
+  (ffi:int8 "wlr_render_texture_with_matrix" (list '* '* '* ffi:float))
+  (not (zero? (% (unwrap-wlr-renderer r)
+                 (unwrap-wlr-texture texture)
+                 (if (>= (length matrix) 9)
+                     (bytestructure->pointer
+                       (bytestructure (bs:vector (length matrix) float)
+                                      (list->vector matrix)))
```
Comment (Z572) [inline wlroots/render/renderer.scm:62]: please allow pass vector

File: wlroots/render/renderer.scm [hunk 7]
```diff
@@ -38,6 +38,19 @@
...
 
+(define-wlr-procedure (wlr-render-texture r texture projection x y alpha)
+  (ffi:int8 "wlr_render_texture" (list '* '* '* ffi:int ffi:int ffi:float))
+  (not (zero? (% (unwrap-wlr-renderer r)
+                 (unwrap-wlr-texture texture)
+                 (if (>= (length projection) 9)
+                     (bytestructure->pointer
+                       (bytestructure (bs:vector (length projection) float)
+                                      (list->vector projection)))
+                   (error "warning: list argument is too small; projection requires, at least, 9"))
```
Comment (Z572) [inline wlroots/render/renderer.scm]: same

File: wlroots/render/renderer.scm [hunk 8]
```diff
@@ -38,6 +38,19 @@
   (ffi:int "wlr_renderer_clear" `(* *))
   (% (unwrap-wlr-renderer renderer) (color->pointer color)))
 
+(define-wlr-procedure (wlr-render-texture r texture projection x y alpha)
+  (ffi:int8 "wlr_render_texture" (list '* '* '* ffi:int ffi:int ffi:float))
+  (not (zero? (% (unwrap-wlr-renderer r)
+                 (unwrap-wlr-texture texture)
+                 (if (>= (length projection) 9)
```
Comment (Z572) [inline wlroots/render/renderer.scm]: same

File: wlroots/types/matrix.scm [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+(define-wlr-procedure (wlr-matrix-project-box box transform rotation projection)
+  (ffi:void "wlr_matrix_project_box" (list '* '* ffi:int32 ffi:float '*))
+  (let* ((max-vector-length 9)
+         (mat (bytestructure (bs:vector max-vector-length float))))
+    (% (bytestructure->pointer mat)
+       (unwrap-wlr-box box)
+       (bs:enum->integer %wl-output-transform-enum transform)
+       rotation
+       (if (>= (length projection) max-vector-length)
```
Comment (Z572) [inline wlroots/types/matrix.scm]: can add a procedure to do convert from 9 length vector or list to bytestructure, and use it.

File: wlroots/types/matrix.scm [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+           (bytestructure->pointer (bytestructure (bs:vector (length projection)
+                                                             float)
+                                                  (list->vector projection)))
```
Comment (Z572) [inline wlroots/types/matrix.scm:21]: same

File: wlroots/types/transform.scm [hunk 11]
```diff
@@ -0,0 +1,13 @@
+(define-module (wlroots types transform)
```
Comment (Z572) [inline wlroots/types/transform.scm]: no need this module, please move procedures to `(wlroots types output)`

File: wlroots/types/transform.scm [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+  #:use-module (wayland protocol)
+  #:use-module (wlroots types)
+  #:use-module (wlroots utils))
+
+(define-wlr-procedure (wlr-output-transform-invert tr)
+  ('* "wlr_output_transform_invert" '(*))
```
Comment (Z572) [inline wlroots/types/transform.scm]: use `ffi:int32` like `(ffi:int32 "wlr_output_transform_invert" (list ffi:int32))`

File: wlroots/types/transform.scm [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  (% (bs:enum->integer %wl-output-transform-enum tr)))
+
+(define-wlr-procedure (wlr-output-transform-compose tr_a tr_b)
+  ('* "wlr_output_transform_compose" '(* *))
```
Comment (Z572) [inline wlroots/types/transform.scm]: same

File: wlroots/types/transform.scm [hunk 14]
```diff
@@ -0,0 +1,13 @@
+(define-module (wlroots types transform)
+  #:use-module (wayland protocol)
+  #:use-module (wlroots types)
+  #:use-module (wlroots utils))
+
+(define-wlr-procedure (wlr-output-transform-invert tr)
+  ('* "wlr_output_transform_invert" '(*))
+  (% (bs:enum->integer %wl-output-transform-enum tr)))
+
+(define-wlr-procedure (wlr-output-transform-compose tr_a tr_b)
```
Comment (Z572) [inline wlroots/types/transform.scm]: tr-a tr-b

File: wlroots.scm [hunk 15]
```diff
@@ -19,6 +19,7 @@
              (types damage-ring)
              (types data-control)
              (types data-device)
+             (types fullscreen-shell-v1)
```
Comment (Z572) [inline wlroots.scm]: move this to add scm file's same commit

File: example/fullscreen-shell.scm [hunk 16]
```diff
@@ -0,0 +1,522 @@
+#!/usr/bin/env -S guile --no-auto-compile -e main
+!#
+;;; An implementation of https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/0.16/examples/fullscreen-shell.c?ref_type=heads
+;; functions retrieved from each module as comments above the module
+(use-modules (ice-9          curried-definitions)
+             ;; format
+             (ice-9          format)
```
Comment (Z572) [inline example/fullscreen-shell.scm:7]: `((ice-9 format) #:select (format))` is better than comment.

File: example/fullscreen-shell.scm [hunk 17]
```diff
@@ -0,0 +1,522 @@
+#!/usr/bin/env -S guile --no-auto-compile -e main
+!#
+;;; An implementation of https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/0.16/examples/fullscreen-shell.c?ref_type=heads
+;; functions retrieved from each module as comments above the module
+(use-modules (ice-9          curried-definitions)
```
Comment (Z572) [inline example/fullscreen-shell.scm:5]: just one space

File: example/fullscreen-shell.scm [hunk 18]
```diff
@@ -0,0 +1,522 @@
...
+             (wlroots        util                box)
+             ;; wlr-log-init
+             (wlroots        util                log))
+
+(default-duplicate-binding-handler
+  '(merge-generics replace warn-override-core warn last))
+
+(define* (add-listen* obj symbol proc #:key (destroy-when         obj)
+                                            (remove-when-destroy? #t))
+  (let ([listener (make-wl-listener proc)])
```
Comment (Z572) [inline example/fullscreen-shell.scm:83]: please don't use `[]`, `()` is good.

File: example/fullscreen-shell.scm [hunk 19]
```diff
@@ -0,0 +1,522 @@
...
+;;; Lines 83–84
+;; struct fullscreen_output *output =
+;; 	wl_container_of(listener, output, frame);
+;
+; (we can skip a call to wl-container-of by just passing in the
+; output we created at the time we were adding this listener)
+(define ((output-handle-frame fullscreen-output) listener data)
+  ;;; Line 85
+  ;; struct wlr_renderer *renderer = output->server->renderer;
+  (let-values ([(renderer)     (get-server-renderer (get-output-server fullscreen-output))]
```
Comment (Z572) [inline example/fullscreen-shell.scm]: use `(srfi srfi-71)`'s `let`

File: wlroots/types/drm.scm [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+            .display-destroy
+            .formats
+            .global
+            .node-name))
+
+;;;
+;; A stub implementation of Mesa's wl_drm protocol.
+;;
+;; It only implements the minimum necessary for modern clients to behave
+;; properly. In particular, flink handles are left unimplemented.
```
Comment (Z572) [inline wlroots/types/drm.scm]: please remove this comment, guile-wlroots is a ffi library

**Review by Z572 — CHANGES_REQUESTED** [review]

**Comment by Z572** [comment]
you can try it at tty use tmux and ssh.
run `example/fullscreen-shell.scm` at tmux, and open a ssh to connect you computer, use `tmux a` to attach tmux, and look log or send <kbd>C-c</kbd>

**Comment by Z572** [comment]
i get a black screen after run it, is normal?

**Comment by 4tuw2tuv0** [comment]
Hmm; I'm not sure I'm doing the `tmux` thing properly. I started `tmux` in one of the TTYs and then went back to my main TTY and ran a `tmux a`, like you suggested. It connected but running `example/fullscreen-shell.scm` caused it to switch from TTY 1 to the other TTY and I was stuck as before, with the screen unable to process input so not able to do anything.

As to getting a black screen, I believe so? The only comment they've provided new users is "A minimal fullscreen-shell server. It only supports rendering." so I think it's very similar to the `scene-graph` example, just not using the new Scene Graph API.

**Comment by Z572** [comment]
i rebase and pushed at 490e276aa1943447a069c6c4e042614dfb259903, but not push example/fullscreen.scm, need check why black screen, 
wait https://github.com/Z572/guile-wlroots/tree/fullscreen-shell-example-fixed.

**Comment by Z572** [comment]
now add at f47d4bd6bdac5e60e6a278e996d9c84d72f6f610, because `wlr-renderer-end` is in `(when (get-output-surface fullscreen-output) something)`, when move out, success.

**Pull request closed without merging**